### PR TITLE
Bug 1254039 - [TV][Home] Move rename, rearrange, and delete from edit mode to context menu.

### PR DESCRIPTION
**Pull request by Fischer-L**
This patch:
- Enable context menu based on app type to edit cards:
  - deck type: move,
  - folder type: move, edit folder,
  - tv, app, appbookmark: move, rename, remove
- Context menu only can be triggered in the normal state
- Remove the edit button so as to enter edit mode only through context menu

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4bc34e7fab6e8240cff959dd15ab5afad9429ee2

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c2cb46c24a928427e917e83ce35203839cc610d6

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c2cb46c24a928427e917e83ce35203839cc610d6

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=80d224da5de4dfed0a37d5d5a46a808c723e261c

File: tv_apps/smart-home/js/home.js [hunk 1]
```diff
@@ -696,18 +853,29 @@
...
     get focusScrollable() {
       return this._focusScrollable;
+    },
+
+    get mode() {
+      return this.mainSection.dataset.mode;
+    },
+
+    set mode(newMode) {
+      if (this.mainSection.dataset.mode != newMode) {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: Unnecessary `if`.

File: tv_apps/smart-home/js/home.js [hunk 2]
```diff
@@ -187,10 +195,148 @@
...
+        case 'tv':
+        case 'app':
+        case 'appbookmark':
+          this.contextmenuElem.appendChild(this.renameMenuItem);
+          this.contextmenuElem.appendChild(this.removeMenuItem);
+          break;
+      }
+    },
+
+    enterEditMode: function (onEnter) {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: Uncertain naming with `toggleEditMode`.

File: tv_apps/smart-home/js/home.js [hunk 3]
```diff
@@ -187,10 +195,148 @@
...
+        this.cleanFolderScrollable();
+        this.edit.toggleEditMode();
+        // XXX: Reset card filter when entering edit mode
+        this.filterManager.resetFilter();
+      });
+    },
 
-        that.editFolderElem.addEventListener(
-                                        'click', that.onEditFolder.bind(that));
+    exitEditMode: function () {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: Uncertain naming with `toggleEditMode`.

File: tv_apps/smart-home/js/home.js [hunk 4]
```diff
@@ -187,10 +195,148 @@
...
+        case 'app':
+        case 'appbookmark':
+          this.contextmenuElem.appendChild(this.renameMenuItem);
+          this.contextmenuElem.appendChild(this.removeMenuItem);
+          break;
+      }
+    },
+
+    enterEditMode: function (onEnter) {
+      return new Promise((resolve, reject) => {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: What's the requirement to use promise here?

File: tv_apps/smart-home/js/home.js [hunk 5]
```diff
@@ -187,10 +195,148 @@
...
+        var scrollable = this.spatialNavigator.getFocusedElement();
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: wait, how do you confirm that the currently focused element is a `scrollable`?

File: tv_apps/smart-home/js/home.js [hunk 6]
```diff
@@ -187,10 +195,148 @@
...
+      this.moveMenuItem.addEventListener('click', () => {
+        this.enterEditMode().then(() => {
+          this.edit.once('arranged', () => this.exitEditMode());
+          this.edit.toggleArrangeMode();
+        });
+      });
+
+      this.renameMenuItem.addEventListener('click', () => {
+        this.enterEditMode()
+          .then(() => {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: Unnecessary line break.

File: tv_apps/smart-home/js/home.js [hunk 7]
```diff
@@ -187,10 +195,148 @@
...
+        this.edit.toggleEditMode();
+        // XXX: Reset card filter when entering edit mode
+        this.filterManager.resetFilter();
+      });
+    },
 
-        that.editFolderElem.addEventListener(
-                                        'click', that.onEditFolder.bind(that));
+    exitEditMode: function () {
+      this.edit.once('exit-edit-mode', () => {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: I guess you can completely remove these end-of-switching-mode-event and just do them synchronously.

File: tv_apps/smart-home/js/edit.js [hunk 8]
```diff
@@ -65,33 +63,34 @@
...
-            // until those actions to be done before focusing editButton.
-            this.spatialNavigator.focus(this.editButton);
-            this.currentNode.classList.remove('focused');
+            // During the process of writeCardlistInCardStore,
+            // editing works is still under going
+            // and card focus would still change.
+            // So we fire 'exit-edit-mode' event
+            // after the process of writeCardlistInCardStore to let
+            // outside able to do following-up works based on final outcomes.
+            this.fire('exit-edit-mode');
```
Comment (rexboy7) [inline tv_apps/smart-home/js/edit.js]: Is this change really required?

File: tv_apps/smart-home/js/home.js (hunk 1 above)
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: But why you need to copy this property to home.js? Is there any benefit to do this than just use this in `edit.js`?

File: tv_apps/smart-home/js/edit.js [hunk 9]
```diff
@@ -409,24 +409,47 @@
...
-      this.modalDialog.open({
-        message: {
-          textL10nId: 'delete-card-alert'
-        },
-        buttonSettings: [
-          {
-            textL10nId: 'cancel',
-            defaultFocus: true,
-            onClick: this.spatialNavigator.focus.bind(this.spatialNavigator)
+      return new Promise((resolve, reject) => {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/edit.js]: I don't know the purpose of changing this to promise.

File: tv_apps/smart-home/js/home.js (hunk 2 above)
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: Actually I don't think adding a pair of enter/edit mode in home.js is a good idea.

File: tv_apps/smart-home/js/home.js [hunk 10]
```diff
@@ -187,10 +195,148 @@
...
+    enterEditMode: function (onEnter) {
+      return new Promise((resolve, reject) => {
+        var scrollable = this.spatialNavigator.getFocusedElement();
+        this._currentEditedCardId = scrollable.currentItem.dataset.cardId;
+
+        if (scrollable === this.folderScrollable) {
+          // Folder would be cleaned then rebuilt when entering edit mode.
+          // So if we are editing card under folder,
+          // we have to find and make card being edited refocused
+          // after folder is rebuilt.
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: Did you confirmed with UX of clean-and-remove behavior?

File: tv_apps/smart-home/js/edit.js [hunk 11]
```diff
@@ -65,33 +63,34 @@
...
+            // So we fire 'exit-edit-mode' event
+            // after the process of writeCardlistInCardStore to let
+            // outside able to do following-up works based on final outcomes.
+            this.fire('exit-edit-mode');
           }.bind(this));
         this._concealPanel(this.currentScrollable, this.currentNode);
-        this.fire('exit-edit-mode');
       } else {
-        this.mainSection.dataset.mode = 'edit';
+        this.mode = 'edit';
```
Comment (rexboy7) [inline tv_apps/smart-home/js/edit.js]: I guess you need to try to fix the following code: find current focusing scrollable and use it as parameter to `setCurrentScrollable`.

And at the end call `realignToReferenceElement` again if we are focusing on `folderScrollable`.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=06f8cfb2ca6a20799575bf3d57e6e933729e17eb

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=17ca2c1434d7c06590d417873277e7913babcc9b

File: shared/js/smart-screen/card_manager.js [hunk 12]
```diff
@@ -126,7 +126,8 @@
         return that._cardStore.saveData('cardList', cardEntries);
       }).then(function() {
         that._asyncSemaphore.p();
-        if (options && options.cleanEmptyFolder) {
+        if (options && options.cleanEmptyFolder &&
+          emptyFolderIndices.length > 0) {
           that.fire('card-removed', emptyFolderIndices);
```
Comment (Fischer-L) [inline shared/js/smart-screen/card_manager.js]: We should fire card-removed event when there are empty folders to remove. Although firing card-removed event when no empty folder would not cause anything to be removed, it would trigger unnecessary focus change.

File: shared/js/smart-screen/scrollable.js [hunk 13]
```diff
@@ -203,7 +203,6 @@
       // Determine initial focus (depends on reversed or not)
       var initNode = this.getItemFromNode(
                      this.getNode(this._isReversed ? this.length - 1 : 0));
-      this.silentFocus(initNode);
```
Comment (Fischer-L) [inline shared/js/smart-screen/scrollable.js]: Using initNode to calculate the translateX is enough. We don't have to silent focus initNode here.
Doing so would cause:
(1) Original focused item lost focus
(2) Since this is a silent focus, it would not trigger style update. So on the UI, the original focused item is still in the focused style but, under the scrollable the current focused item has became the initNode.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6b50695dc404933e6662858a73a20df60716ad71

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=439090e4985b188c97c1c2f06135a1974244b628

File: tv_apps/smart-home/js/home.js (hunk 1 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: This is because we have to handle mode changing to 'cardPicker' or 'filter' in the home.js as well.

File: tv_apps/smart-home/js/home.js [hunk 14]
```diff
@@ -187,12 +193,77 @@
...
+        this.edit.renameCard(focus, focus.getNodeFromItem(focus.currentItem));
+      });
+
+      this.removeMenuItem.addEventListener('click', () => {
+        var focus = this.spatialNavigator.getFocusedElement();
+        this.edit.toggleEditMode();
+        this.edit.once('delete-card-ended', () => this.edit.toggleEditMode());
+        this.edit.openDeleteCardDialog(focus,
+          focus.getNodeFromItem(focus.currentItem));
+      });
```
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: We have to toggle off the edit mode after deleting-card job ends.
Because the open of dialog is non-blocking so if we do the following code:
```
this.edit.toggleEditMode();
this.edit.openDeleteCardDialog(……..);
this.edit.toggleEditMode();
```
2 issues could happen:
(1) We would exit the edit mode immediately while the deleting-card dialog still remains opened.
(2) If we exit the edit mode before deleting card transition finishes, during exiting the edit mode, folder would still be considered not empty so empty folder would not be removed after exiting the edit mode

File: tv_apps/smart-home/js/edit.js (hunk 9 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/edit.js]: This is to let outside caller know the right timing that deleting card job ends.
The code has been changed to use the event to reduce the complexity.

File: shared/js/smart-screen/card_manager.js (hunk 12 above)
Comment (rexboy7) [inline shared/js/smart-screen/card_manager.js]: should or shouldn't?

File: shared/js/smart-screen/scrollable.js (hunk 13 above)
Comment (rexboy7) [inline shared/js/smart-screen/scrollable.js]: The purpose of this line is, when we show the list reversely, the rightmost item should be focused.
And deleting this would break this feature.

File: tv_apps/smart-home/js/home.js [hunk 15]
```diff
@@ -187,12 +193,77 @@
           cardManager: that.cardManager
         });
         that._cardPicker.on('hide', that.onCardPickerHide.bind(that));
+        that._cardPicker.on('show', that.onCardPickerShow.bind(that));
+      });
+    },
+
+    initContextMenu: function () {
+      this.moveMenuItem.addEventListener('click', () => {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: I'd prefer to use `Home.onMoveMenuItemClick`

File: tv_apps/smart-home/js/home.js [hunk 16]
```diff
@@ -187,12 +193,77 @@
...
+          this.edit.toggleEditMode();
+          this.edit.once('arranged', () => this.edit.toggleEditMode());
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: I'd prefer add this in `edit.toggleArrangeMode` rather than binding it every time

File: tv_apps/smart-home/js/home.js [hunk 17]
```diff
@@ -187,12 +193,77 @@
...
+          this.edit.toggleArrangeMode();
+      });
+
+      this.renameMenuItem.addEventListener('click', () => {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: same as above

File: tv_apps/smart-home/js/home.js [hunk 18]
```diff
@@ -187,12 +193,77 @@
...
+        var focus = this.spatialNavigator.getFocusedElement();
+        this.edit.toggleEditMode();
+        this.edit.once('rename-card-ended', () => this.edit.toggleEditMode());
+        this.edit.renameCard(focus, focus.getNodeFromItem(focus.currentItem));
+      });
+
+      this.removeMenuItem.addEventListener('click', () => {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: same as above.

File: tv_apps/smart-home/js/home.js [hunk 19]
```diff
@@ -187,12 +193,77 @@
...
+    initContextMenu: function () {
+      this.moveMenuItem.addEventListener('click', () => {
+          this.edit.toggleEditMode();
+          this.edit.once('arranged', () => this.edit.toggleEditMode());
+          this.edit.toggleArrangeMode();
+      });
+
+      this.renameMenuItem.addEventListener('click', () => {
+        var focus = this.spatialNavigator.getFocusedElement();
+        this.edit.toggleEditMode();
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: I think calling rename doesn't depend on mode changing, so you just need one line.

File: tv_apps/smart-home/js/home.js (hunk 1 above)
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: Can't they be handled without adding additional modes?

File: tv_apps/smart-home/js/home.js (hunk 14 above)
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: You don't need to switch to edit mode to do card deletion.

File: tv_apps/smart-home/js/home.js [hunk 20]
```diff
@@ -187,12 +193,77 @@
...
+        this.edit.once('rename-card-ended', () => this.edit.toggleEditMode());
+        this.edit.renameCard(focus, focus.getNodeFromItem(focus.currentItem));
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: `this.edit.renameCard(this.focusScrollable, this.focusElem)`

File: tv_apps/smart-home/js/home.js (hunk 14 above)
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: And actually switching to edit mode creates unnecessary animation. Can you try not to switch into edit mode and just remove card directly? Maybe some edit-mode specified css animation rules need to be changed to trigger animationEnd correctly.

File: tv_apps/smart-home/js/home.js [hunk 21]
```diff
@@ -657,12 +725,25 @@
...
+    onFilterChanged: function (filterName) {
+      if (filterName === FilterManager.FILTERS.ALL.name) {
+        this.mode = '';
+      } else {
+        this.mode = 'filter';
+      }
+    },
+
+    onCardPickerShow: function() {
+      this.mode = 'cardPicker';
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: Is this mode really necessary?

File: tv_apps/smart-home/js/home.js [hunk 22]
```diff
@@ -187,12 +193,77 @@
...
-                                        'click', that.onEditFolder.bind(that));
+      this.editFolderMenuItem.addEventListener('click', () => {
+        this._cardPicker.show(this.cardScrollable.currentItem);
       });
     },
 
+    enableContextMenu: function () {
+      var focus = this.spatialNavigator.getFocusedElement();
+      if (this.mode === '' &&
+          (focus === this.cardScrollable || focus === this.folderScrollable)) {
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: I think binding context menu on just `folder-list-frame` and `card-list-frame` could be easier, and you don't need to determine everything for switching contextmenu on and off.

But we have some bad layout that prevents us for doing this. So you may need to add some css rule before doing that:
```
#search-bar, #folder-viewer, #card-viewer {
  pointer-events: none;
}

#card-list-frame, #folder-list-frame {
  pointer-events: auto;
}
```

Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: And after that you just need to disable `contextmenu` when switching filter I think.

File: tv_apps/smart-home/js/edit.js [hunk 23]
```diff
@@ -65,67 +63,75 @@
...
+      } else if (this.mode == 'arrange') {
+        this.mode = 'edit';
 
         this.currentNode = this.currentScrollable.getNodeFromItem(
                                             this.currentScrollable.currentItem);
         this._revealPanel(this.currentScrollable, this.currentNode);
         this._clearHintArrow();
         this.cardScrollable.setColspanOnFocus(0);
         this.folderScrollable.realignToReferenceElement();
+        this.fire('arranged');
```
Comment (rexboy7) [inline tv_apps/smart-home/js/edit.js]: Can be removed due to comments in https://github.com/mozilla-b2g/gaia/pull/34247/files#diff-5a33a94cd737c2a136a3c3139f952015R203.

File: tv_apps/smart-home/js/edit.js [hunk 24]
```diff
@@ -400,11 +406,11 @@
...
         var result = prompt(val, oldName);
-        if (!result) {
-          return;
+        if (result) {
+          card.name = {raw: result};
+          this.cardManager.updateCard(card);
         }
-        card.name = {raw: result};
-        this.cardManager.updateCard(card);
+        this.fire('rename-card-ended');
```
Comment (rexboy7) [inline tv_apps/smart-home/js/edit.js]: This can be removed too.

File: tv_apps/smart-home/js/edit.js [hunk 25]
```diff
@@ -418,15 +424,32 @@
           {
             textL10nId: 'cancel',
             defaultFocus: true,
-            onClick: this.spatialNavigator.focus.bind(this.spatialNavigator)
+            onClick: () => {
+              this.spatialNavigator.focus();
+              this.fire('delete-card-ended');
```
Comment (rexboy7) [inline tv_apps/smart-home/js/edit.js]: Should be an unnecessary event.

File: shared/js/smart-screen/card_manager.js (hunk 12 above)
Comment (Fischer-L) [inline shared/js/smart-screen/card_manager.js]: Should ONLY fire card-removed event when there are empty folders.
Should NOT fire card-removed event when there is no empty folders.

File: shared/js/smart-screen/scrollable.js (hunk 13 above)
Comment (Fischer-L) [inline shared/js/smart-screen/scrollable.js]: This is for the case that when toggling edit mode on, we have to realign reference element and then trigger this line of calling silentFocus. Since this silentFocus call has the reason to exist. I will handle refocus back to the original focused item in the toggleEditMode method.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2661b425581a2648e1a6847f412341de938c3ec8

File: tv_apps/smart-home/js/home.js (hunk 14 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: OK, updated to not switching into edit mode and just remove card directly.

File: tv_apps/smart-home/js/home.js (hunk 15 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: OK, added onMoveMenuItemClick method

File: tv_apps/smart-home/js/home.js (hunk 16 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: Updated to do this in edit.toggleArrangeMode

File: tv_apps/smart-home/js/home.js (hunk 17 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: Added onRenameMenuItemClick method

File: tv_apps/smart-home/js/home.js (hunk 18 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: Added onRemoveMenuItemClick method

File: tv_apps/smart-home/js/home.js (hunk 19 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: OK, updated to not switching into edit mode for renaming.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=53d20cda3a13deb146dfa5a4469658966898a674

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ff39c0e68e003c593a66a0e5618b03aae704c660

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4fbb0f7191f9331b901136cc4b0ecc8beecf9727

File: tv_apps/smart-home/js/home.js (hunk 22 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: Updated to bind context menu to card-list and folder-list. This requires to handle css z-index to enable triggering contextmenu event on card-list and folder-list but eliminate the handling of cardPicker mode.

File: tv_apps/smart-home/js/home.js (hunk 21 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: Updated to to bind context menu to card-list and folder-list so this mode is eliminated.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=af392c796e6f43815258eb83eb4db507f02793b2

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=91699def59778b194ad69bea812cb05428c3b331

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=674f1cebf296ca84ade8ee6184b2f0f6ad296cd7

File: tv_apps/smart-home/js/edit.js [hunk 26]
```diff
@@ -65,67 +63,78 @@
       this.isFolderReady = false;
       // The hovering card
       this._hoveringCard = null;
-      this.mainSection.dataset.mode = '';
+      this.mode = '';
```
Comment (luke-chang) [inline tv_apps/smart-home/js/edit.js]: Let's use `this._home.mode` directly and remove the getter and setter of `Edit` object.

File: tv_apps/smart-home/js/home.js [hunk 27]
```diff
@@ -187,12 +193,96 @@
...
         that._cardPicker.on('hide', that.onCardPickerHide.bind(that));
-
-        that.editFolderElem.addEventListener(
-                                        'click', that.onEditFolder.bind(that));
       });
     },
 
+    initContextMenu: function () {
+      this.moveMenuItem.addEventListener('click',
+        () => this.onMoveMenuItemClick());
```
Comment (luke-chang) [inline tv_apps/smart-home/js/home.js]: Let's use `.bind()` here to align with others.

File: tv_apps/smart-home/js/home.js [hunk 28]
```diff
@@ -187,12 +193,96 @@
...
+
+    enableContextMenu: function () {
+      if (this.mode !== '') {
+        return;
+      }
+      var focus = this.spatialNavigator.getFocusedElement();
+      if (focus === this.cardScrollable) {
+        this.cardListElem.setAttribute('contextmenu', this.contextmenuElem.id);
+        this.folderListElem.setAttribute('contextmenu', '');
+        this.cardListElem.style.zIndex = '1';
```
Comment (luke-chang) [inline tv_apps/smart-home/js/home.js]: As offline discussion, let's consider how to fix `z-index` issue.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=15102eed2ce69b9379b82ed4a87585079a726154

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=dc3a355b4a274562c527ebd9b762ef586c14b5ec

File: tv_apps/smart-home/js/home.js (hunk 27 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: OK, updated.

File: tv_apps/smart-home/js/home.js (hunk 28 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/home.js]: Updated to unify the controlling of enabling/disabling context menu through modes, thank you

File: tv_apps/smart-home/js/home.js [hunk 29]
```diff
@@ -657,12 +731,25 @@
...
+    onFilterChanged: function (filterName) {
+      if (filterName === FilterManager.FILTERS.ALL.name) {
+        this.mode = '';
+      } else {
+        this.mode = 'filter';
+      }
+    },
+
+    onCardPickerShow: function () {
+      this.mode = 'cardPicker';
```
Comment (luke-chang) [inline tv_apps/smart-home/js/home.js]: `card-picker`

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fdd7dd61bbc9525c8c12aac7c236855d0cab9eb6
